pkg/ring/lifecycler.go

```diff
@@ -556,21 +559,13 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 		// but we need to update the local state accordingly.
 		i.setRegisteredAt(instanceDesc.GetRegisteredAt())
 
-		// If the ingester is in the JOINING state this means it crashed due to
-		// a failed token transfer or some other reason during startup. We want
-		// to set it back to PENDING in order to start the lifecycle from the
-		// beginning.
-		if instanceDesc.State == JOINING {
-			level.Warn(log.Logger).Log("msg", "instance found in ring as JOINING, setting to PENDING",
-				"ring", i.RingName)
+		// The instance may already be in the ring with one of the following states:
+		// - JOINING: this means it crashed due to a failed token transfer or some other reason during startup.
+		// - LEAVING: this means it crashed while shutting down and unregister from ring on shutdown is disabled.
+		// In both cases, we want to set it back to PENDING in order to start the lifecycle from the beginning.
+		if instanceDesc.State == JOINING || instanceDesc.State == LEAVING {
+			level.Info(log.Logger).Log("msg", "instance already found in ring, changing state to PENDING", "state", instanceDesc.State, "ring", i.RingName)
 			instanceDesc.State = PENDING
-			return ringDesc, true, nil
-		}
-
-		// If the ingester failed to clean it's ring entry up in can leave it's state in LEAVING.
-		// Move it into ACTIVE to ensure the ingester joins the ring.
-		if instanceDesc.State == LEAVING && len(instanceDesc.Tokens) == i.cfg.NumTokens {
-			instanceDesc.State = ACTIVE
```

Comment:
This was done in https://github.com/cortexproject/cortex/pull/1921. It was updating `instanceDesc`, but the following `return nil, true, nil` was not really updating the ring. The reason why it was working was because the `instanceDesc.State` was set on the local state and then updated during the next heartbeat.

The new logic in this PR `return ringDesc, true, nil` if `instanceDesc` has changed, which I believe is easier to understand and less error prone.